server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java

```diff
@@ -195,6 +196,16 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
                         currentNodeCount,
                         maximumNodeCount);
             }
+
+            if (currentNodeCount == maximumNodeCount && (shardCount % numberOfAttributes) != 0) {
+                for (ObjectCursor<String> cursor : nodesPerAttribute.keys()) {
+                    if (shardPerAttribute.get(cursor.value) == 0) {
```

Comment:
This seems wrong and I think would in your example result in always returning `NO`.
In your example for `rack`: `maximumNodeCount == 1` and currentNodeCount will also always be `1` in the initial iteration => we always get `NO` because there's always going to be an element in `shardPerAttribute` that is `0`?

Reply:
@original-brownbear Thanks review this PR.
I've tested many scenarios, including restarting clusters, index closes etc. 
You're worried that this won't happen, because if maximumNodeCount = 1, (shardCount %  numberOfAttributes) must be equal to 0.

Reply:
@hanbj 

>You're worried that this won't happen, because if maximumNodeCount = 1, (shardCount % numberOfAttributes) must be equal to 0.

I don't think so and there's even an example of that in your PR description with the racks:
* You have `8` values for `rack` => `maximumNode == 1` for `5` shards but `shardCount % numberOfAttributes` is not `0` and you will always have `1` for `currentNodeCount` when testing a possible configuration?

Reply:
@hanbj as a matter of fact, with this PR all tests fail for that very reason as far as I can see. Just run `./gradlew :server:test` and you should see the resulting failures.

Reply:
@original-brownbear Thanks, this is too bad, I used the wrong remainder operation, I have modified it.

Reply:
@hanbj thanks this looks better and passes tests now :) 

Two points remaining now:

1. We need a test for this. Would you want to work on one? We should do some randomized testing that ensures that this works properly for any random combination of number of attributes and replicas.

Leading me to my next point:

2. I'm not sure this works correctly for cases where `replicaCount + 1 > 2 * numberOfValuesInAttribute`. E.g. assume we have `4` values for `switch` but now we have 9 replicas (including the primary): We should make sure that in that case we get a distribution of `2` replicas for `3` values and  `4` replicas on one value. Not something like `4` replicas on one value, `3` on another one and then `1` on each of the remaining two values right?
It seems like the current code does not ensure that, but it should be easy to extend it to do so once we have a test.